src/librustc_trans/trans/mir/block.rs

```diff
@@ -105,65 +96,140 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
+                    match self.trans_operand(bcx, arg).val {
                         Ref(llval) | Immediate(llval) => llargs.push(llval),
-                        FatPtr(base, extra) => {
-                            // The two words in a fat ptr are passed separately
-                            llargs.push(base);
-                            llargs.push(extra);
+                        FatPtr(b, e) => {
+                            llargs.push(b);
+                            llargs.push(e);
                         }
                     }
                 }
 
-                // FIXME: Handle panics
-                //let panic_bb = self.llblock(targets.1);
-                //self.make_landing_pad(panic_bb);
-
-                // Do the actual call.
-                let (llret, b) = base::invoke(bcx,
-                                              callee.immediate(),
-                                              &llargs[..],
-                                              callee.ty,
-                                              DebugLoc::None);
-                bcx = b;
-
-                // Copy the return value into the destination.
-                if let ty::FnConverging(ret_ty) = ret_ty {
-                    if !type_of::return_uses_outptr(bcx.ccx(), ret_ty) &&
-                       !common::type_is_zero_size(bcx.ccx(), ret_ty) {
-                        base::store_ty(bcx, llret, call_dest.llval, ret_ty);
+                // Many different ways to call a function handled here
+                match (base::avoid_invoke(bcx), kind) {
+                    // The two cases below are the only ones to use LLVM’s `invoke`.
+                    (false, &mir::CallKind::DivergingCleanup(cleanup)) => {
+                        let cleanup = self.bcx(cleanup);
+                        let landingpad = self.make_landing_pad(cleanup);
+                        build::Invoke(bcx,
+                                      callee.immediate(),
+                                      &llargs[..],
+                                      self.unreachable_block().llbb,
+                                      landingpad.llbb,
+                                      Some(attrs),
+                                      debugloc);
+                    },
+                    (false, &mir::CallKind::ConvergingCleanup { ref targets, .. }) => {
+                        let cleanup = self.bcx(targets.1);
+                        let landingpad = self.make_landing_pad(cleanup);
+                        build::Invoke(bcx,
+                                      callee.immediate(),
+                                      &llargs[..],
+                                      self.llblock(targets.0),
+                                      landingpad.llbb,
+                                      Some(attrs),
+                                      debugloc);
+                        if must_copy_dest {
+                            // FIXME: What do we do here?
```

Comment:
The problem here (is much harder than it looks at the first look!) is that `invoke` is a basic block terminator and thus we cannot add anything else to the current block. We would have to add, or rather, *prepend* a store instruction to a successful branch, but… that’s tricky, because:

1. That’s hacky (duh);
2. I can’t seem to see anyway to do it;
3. It would make some already-generated blocks invalid (e.g. we would have to assert that success branch has higher basic-block number).

Basically, to me it seems like we should change the MIR some more and figure out if destination must be copied (or rather, avoid copying destination altogether) during the MIR build phase.